### PR TITLE
AYR-1842 - Refactor browse/ route to use a main logic processing function or class

### PR DESCRIPTION
**Pull request by karwan-zaizi**
## JIRA ticket
[AYR-1842](https://national-archives.atlassian.net/browse/AYR-1842)

### After

- [ ] Requires env variable(s) to be updated

[AYR-1842]: https://national-archives.atlassian.net/browse/AYR-1842?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/main/routes.py [hunk 1]
```diff
@@ -175,84 +256,11 @@ def accessibility():
...
-            pagination=pagination,
-            filters=filters,
-            date_filters=date_filters,
-            sorting_orders=sorting_orders,
-            num_records_found=num_records_found,
-            query_string_parameters=validated_data,
-            id=None,
-        )
+    validated_browse_data = BrowseRoute(request.validated_data)
+    result = validated_browse_data.get_browse_route_data(form)
```
Comment (anthonyhashemi) [inline app/main/routes.py]: Why do we need these 2 returns, we just need `return render_template(` as before

Update: Ah its because of the redirect you have in the processing method, but yeah if we move that out we won't need that down here

File: app/main/routes.py [hunk 2]
```diff
@@ -175,84 +256,11 @@ def accessibility():
...
-            transferring_bodies=transferring_bodies,
-            pagination=pagination,
-            filters=filters,
-            date_filters=date_filters,
-            sorting_orders=sorting_orders,
-            num_records_found=num_records_found,
-            query_string_parameters=validated_data,
-            id=None,
-        )
+    validated_browse_data = BrowseRoute(request.validated_data)
```
Comment (anthonyhashemi) [inline app/main/routes.py]: The route looks much simpler to parse now! 

I would suggest that it could be better to keep the authorization check and redirect logic in the route itself and have the class or function just handle processing.

This way the route decides who can access what and which endpoint they should use, which is nice and explicit and the processing function just decides what to do with the data.

 e.g. something like

```
def browse():
    """
    Render the browse page for all-access users.

    Standard users are redirected to their specific transferring body page.
    All-access users can browse all transferring bodies with filtering and sorting.

    Returns:
        For standard users: Redirect to their transferring body browse page
        For all-access users: Rendered browse.html template with results
    """
    form = SearchForm()

    # Check user permissions and redirect standard users
    ayr_user = AYRUser(session.get("user_groups"))
    if ayr_user.is_standard_user:
        return redirect(
            f"/browse/transferring_body/{ayr_user.transferring_body.BodyId}"
        )

    # All-access user: gather transferring bodies
    transferring_bodies = [body.Name for body in Body.query.all()]

    # Get validated request data
    validated_data = request.validated_data

    # Process the browse request (business logic)
    result_data = process_browse_request(
        validated_data=validated_data,
        default_page_size=int(current_app.config["DEFAULT_PAGE_SIZE"]),
        transferring_bodies=transferring_bodies,
    )

    # Render template with the processed data
    return render_template(
        "browse.html",
        form=form,
        browse_type="browse",
        id=None,
        **result_data,
    )

```

Also I have just used a function in this snippet, as I couldn't see what benefit we'd get from a class in this scenario with slightly more boilerplate. It might be useful in the future, but for now I feel if it can be kept simpler, we don't need to prematurely abstract it.



Also worth moving the new processing logic out into a new folder/file

**Comment by ghost** [comment]
### Wiz Scan Summary

| Scanner | Findings |
|:--------|---------:|
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/vulnerability_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/vulnerability_light.svg">Vulnerability Finding</picture></a>&ensp;Vulnerabilities | - |
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/data_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/data_light.svg">Data Finding</picture></a>&ensp;Sensitive Data | - |
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/secret_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/secret_light.svg">Secret Finding</picture></a>&ensp;Secrets | - |
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/configuration_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/configuration_light.svg">IaC Misconfiguration</picture></a>&ensp;IaC Misconfigurations | - |
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/sast_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/sast_light.svg">SAST Finding</picture></a>&ensp;SAST Findings | 1 <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_light.svg">Medium</picture></a> |
| <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/supplyChain_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/finding_type_icons/supplyChain_light.svg">Software Management Finding</picture></a>&ensp;Software Management Findings | - |
| | |
| **Total** | 1 <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_light.svg">Medium</picture></a> |

*<a href="https://app.wiz.io/findings/code-cicd-scans#~(event~(~%2700699c3d-900b-8107-b299-3cbcf9df87f4*2cCI_CD_SCAN*2c2026-02-23T11*3a44*3a16.402001846Z))" target="_blank">View scan details in Wiz</a>*

To detect these findings earlier in the dev lifecycle, try using *<a href="https://marketplace.visualstudio.com/items?itemName=WizCloud.wiz-vscode" target="_blank">Wiz Code VS Code Extension.</a>*
### Pull Request Developer Guidance
If the link to the Wiz scan details doesn't work, please ensure you are using the right role in Wiz and that the repository is assigned to the correct project within our Wiz Terraform. Or, you can ask for help in #tna-cloud-security on Slack.

File: app/main/process_routes/browse_route.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        sorting_orders["transferring_body"] = "asc"
+    query = build_browse_query(
+        filters=filters,
+        sorting_orders=sorting_orders,
+    )
+    browse_results = query.paginate(
+        page=page, per_page=per_page
+    )
+    total_records = db.session.query(
+        func.sum(query.subquery().c.records_held)
```
Comment (ghost) [inline app/main/process_routes/browse_route.py:63]: <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/long_severity_tags/medium_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/long_severity_tags/medium_light.svg">Medium</picture></a>&ensp;SAST Finding

**SQL Injection in Manually Constructed Queries** (CWE-89)

<details>
<summary>More Details</summary>
<br>


SQL injection is a technique where malicious SQL statements are inserted into application-generated queries to manipulate the database. This vulnerability arises when user input is directly concatenated into SQL queries without proper sanitization. An attacker can craft input that modifies the intended logic of the query, potentially allowing unauthorized data access, modification, or deletion. The consequences can be severe, including data breaches, data corruption, and complete system compromise. To avoid SQL injection, always use parameterized queries or prepared statements that separate user input from the query logic. Never construct dynamic SQL queries by concatenating user input.
<br>

| Attribute | Value |
| ----------- | -------- |
| **Impact** | <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_light.svg">Medium</picture></a> |
| **Likelihood** | <a><picture><source media="(prefers-color-scheme: dark)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://assets.wiz.io/wiz-code/short_severity_tags/medium_light.svg">Medium</picture></a> |


### Remediation

SQL injection is a technique where malicious SQL statements are inserted into application-generated queries to manipulate the database. This can lead to unauthorized data access, data loss, and even full system compromise. SQL injection attacks are a serious threat to web applications that interact with databases.

To remediate SQL injection vulnerabilities, user input should never be directly concatenated into SQL queries. Instead, use parameterized queries or prepared statements provided by your database library. These mechanisms automatically escape user input and prevent it from being interpreted as part of the SQL query.

Code examples:

// VULNERABLE CODE - User input is concatenated into the SQL query
query = "SELECT * FROM users WHERE name = '" + user_input + "'"
cursor.execute(query)

// SECURE CODE - User input is passed as a parameter to the query
query = "SELECT * FROM users WHERE name = ?"
cursor.execute(query, (user_input,))

Additional recommendations:
- Follow the principle of least privilege and only grant the minimum required database permissions.
- Use the latest versions of database libraries and frameworks, as they often include security improvements.
- Implement input validation and sanitization to filter out malicious input before it reaches the database layer.
- Regularly review and test your application for SQL injection vulnerabilities, especially after making changes or updates.
- Adhere to secure coding practices outlined in standards like the OWASP Top 10 and the CERT Secure Coding Standards.
- Consider using an Object-Relational Mapping (ORM) library, which can provide an additional layer of abstraction and security when interacting with databases.
- As an alternative approach, consider using NoSQL databases or other data storage solutions that are less susceptible to SQL injection attacks.


> Rule ID: WS-PYTHON-00330
<br>

To ignore this finding as an exception, reply to this conversation with `#wiz_ignore reason`


If you'd like to ignore this finding in all future scans, add an exception in the .wiz file ([learn more](https://docs.wiz.io/wiz-docs/docs/exception-management-with-wiz-code#repository-configuration-file-wiz)) or create an Ignore Rule ([learn more](https://docs.wiz.io/wiz-docs/docs/exception-management-with-wiz-code#ignore-rules)).
</details>
<br>

To get more details on how to remediate this issue using AI, reply to this conversation with `#wiz remediate`

File: app/main/routes.py (hunk 1 above)
Comment (karwan-zaizi) [reply app/main/routes.py]: @anthonyhashemi This is resolved

File: app/main/routes.py (hunk 2 above)
Comment (karwan-zaizi) [reply app/main/routes.py]: @anthonyhashemi this is ready for another review

File: app/main/process_routes/browse_route.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+from app.main.db.queries import build_browse_query
+from app.main.util.date_filters_validator import validate_date_filters
+from app.main.util.filter_sort_builder import (
+    build_filters,
+    build_sorting_orders,
+)
+from app.main.util.pagination import get_pagination
+
+
+def process_browse_request(
```
Comment (anthonyhashemi) [inline app/main/process_routes/browse_route.py:16]: We should tests this function now. We should be able to move most tests from the endpoint level down to this level now. The endpoint tests will be much simpler as it shouldn't care about the edge cases handled by this anymore

File: app/main/routes.py (hunk 1 above)
Comment (anthonyhashemi) [reply app/main/routes.py]: amazing, thanks for addressing

File: app/main/routes.py (hunk 2 above)
Comment (anthonyhashemi) [reply app/main/routes.py]: cheers, have reviewed and left one comment about tests

File: app/tests/test_browse_route.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
+
```
Comment (anthonyhashemi) [inline app/tests/test_browse_route.py]: The test file path seems a bit misleading, I would think that is testing the browse route itself, not the processing function.

Also I suggest that the equivalent tests tested here, can now be removed from the route level tests for browse as the code paths would be covered here. It is still worth one success and one failure to test at endpoint level alongside any other logic that remains there like validation etc.

Let me know if you want me to clarify at all.

Highly recommend https://martinfowler.com/articles/practical-test-pyramid.html which goes deeper into this testing philosophy.

2 quotes pertinent to this:


"**_As with production code you should strive for simplicity and avoid duplication. In the context of implementing your test pyramid you should keep two rules of thumb in mind:
If a higher-level test spots an error and there's no lower-level test failing, you need to write a lower-level test
Push your tests as far down the test pyramid as you can_**"


and



"**_second rule is important to keep your test suite fast. If you have tested all conditions confidently on a lower-level test, there's no need to keep a higher-level test in your test suite. It just doesn't add more confidence that everything's working. Having redundant tests will become annoying in your daily work. Your test suite will be slower and you need to change more tests when you change the behaviour of your code._**"

**Comment by colinbowen** [comment]
PR looks good to me, it could still do with testing the process_browse_request function that calls the function directly not via HTTP then asserts on the returned dictionary values

File: app/tests/test_process_browse_route.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+            """
+            Given an all_access_user accessing the browse page
+            When they make a GET request with an invalid page value (e.g., page=0 or page=-1 or page=abc)
+            Then they should be redirected to the default page (page 1)
+            """
+            mock_all_access_user(client)
+            app.config["DEFAULT_PAGE_SIZE"] = 2
+
+            for invalid_page in ["-1", "0", "abc"]:
+                response = client.get(f"{self.route_url}?page={invalid_page}", follow_redirects=True)
```
Comment (anthonyhashemi) [inline app/tests/test_process_browse_route.py]: we should avoid for loops in tests; each test should test a specific scenario so it is a good piece of documentation and easier to debug/maintain. 
Split these out; or maybe one of these 3 scenarios is enough given we have page validation tests eslewhere

File: app/tests/test_process_browse_route.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+from app.main.process_routes.browse_route import process_browse_request
+from app.tests.test_browse import (
+    TestBrowse,
+)
+
+
+class TestProcessBrowseRoute(TestBrowse):
+
+    def test_browse_filter_no_results(
+        self, client: FlaskClient, mock_all_access_user
```
Comment (colinbowen) [inline app/tests/test_process_browse_route.py]: These tests are exactly what we needed. It would be good to expand them to include the different available sort options we have e.g.: 

- Filter by transferring_body_filter
- Filter by series_filter
- Filter by a date range
- Sort by different fields
- one with a date validation error to check error state

File: app/main/process_routes/browse_route.py (hunk 4 above)
Comment (karwan-zaizi) [reply app/main/process_routes/browse_route.py:16]: Updated the tests

File: app/tests/test_process_browse_route.py (hunk 6 above)
Comment (karwan-zaizi) [reply app/tests/test_process_browse_route.py]: This has been updated

File: app/tests/test_process_browse_route.py (hunk 7 above)
Comment (karwan-zaizi) [reply app/tests/test_process_browse_route.py]: @colinbowen These test cases need to wait till we updated the other routes, this ticket is only for browse route

File: app/tests/test_process_browse_route.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
+from flask import Response
+from flask.testing import FlaskClient
+
+from app.main.process_routes.browse_route import process_browse_request
+from app.tests.test_browse import (
```
Comment (colinbowen) [inline app/tests/test_process_browse_route.py]: Also, the tests are running twice because TestProcessBrowseRoute inherits from TestBrowse.

You can do this which should work
```suggestion
from app.main.process_routes.browse_route import process_browse_request


class TestProcessBrowseRoute:

    @property
    def route_url(self):
        return "/browse"

```

File: app/tests/test_process_browse_route.py (hunk 7 above)
Comment (colinbowen) [reply app/tests/test_process_browse_route.py]: We do have different filter options available for browse, which would be good to test as part of these changes. I believe one for each would be enough to cover the available options.

Screenshot 2026-02-16 at 15 07 54

Comment (colinbowen) [reply app/tests/test_process_browse_route.py]: Tests look much better. Good work

File: app/tests/test_browse_route.py (hunk 5 above)
Comment (colinbowen) [reply app/tests/test_browse_route.py]: Tests has been updated to test against the process_browse_route function and not against the routes.

**Review by colinbowen — APPROVED** [review]
